Connect API with migration algorithm

jonathanknowles opened the pull request:
# Issue Number

ADP-840

# Overview

This PR:

- [x] Implements `Api.Server.createWalletMigrationPlan`.
- [x] Implements `Api.Server.migrateWallet`.
- [x] Adjusts `migrateWallet` to require a non-empty list of addresses.
- [x] Adjusts the success response types for all migration endpoints to be `202` `ACCEPTED`.
- [x] Resurrects all previously-disabled integration tests (for ada-specific wallets).
- [x] Removes the "disabled" warnings on migration endpoints.
- [x] Adds integration test coverage for MA wallets.

# QA Due Diligence

After rewriting the integration test suite, I ran the entire integration test suite 500 times in an effort to increase confidence that there are no flaky tests. The test suite passed 500 times, with no failures or errors.

jonathanknowles committed:
Add function `Wallet.migrationPlanToUnsignedTxs`.

jonathanknowles committed:
Add function `Wallet.createMigrationPlan`.

jonathanknowles committed:
Implement function `Api.Server.createMigrationPlan`.

jonathanknowles committed:
Make `ApiWalletMigrationPostData.addresses` non-empty.

Supplying an empty list of addresses would make it impossible to perform
a migration.

jonathanknowles committed:
Implement function `Api.Server.migrateWallet`.

jonathanknowles committed:
Decompose function `Wallet.migrationPlanToUnsignedTxs`.

The responsibility for turning a `SelectionResult` and a `Withdrawal`
into an `UnsignedTx` can be located outside of this function.

In fact, it is advantageous to do so, since signing a transaction with
`signTransaction` requires a `SelectionResult`, and not a `UnsignedTx`.

jonathanknowles committed:
Decompose function `Wallet.signTransaction`.

This function had the dual responsibility of assigning change addresses
and signing the resultant transaction.

However, we sometimes need to sign transactions for which change addresses
have already been assigned, or transactions without any change. In the
second case, it's overly restrictive to require that callers provide a
way of making change addresses.

Therefore, it makes sense to decouple these two behaviours.

This change extracts out a new function `assignChangeAddressesAndUpdateDb`,
which handles the responsibility of assigning change addresses and
updating the DB.

jonathanknowles committed:
Use correct response types for migration endpoints.

The integration tests expect these to be 202 ACCEPTED.

jonathanknowles committed:
Resurrect test `CREATE_MIGRATION_PLAN_01`.

jonathanknowles committed:
Resurrect test `CREATE_MIGRATION_PLAN_02`.

jonathanknowles committed:
Resurrect test `CREATE_MIGRATION_PLAN_03`.

jonathanknowles committed:
Resurrect test `CREATE_MIGRATION_PLAN_04`.

jonathanknowles committed:
Resurrect test `CREATE_MIGRATION_PLAN_05`.

jonathanknowles committed:
Resurrect test `MIGRATE_01`.

jonathanknowles committed:
Resurrect test `MIGRATE_02`.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs:
```diff
@@ -193,7 +220,7 @@ spec = describe "SHELLEY_MIGRATIONS" $ do
               testAddressCycling 3
               testAddressCycling 10
 
-    Hspec.it "SHELLEY_MIGRATE_01_big_wallet - \
```
🙅‍♂️ I would not. `Hspec.it` doesn't retry on failures. `it` does retry, but the funds of the hard-coded mnemonic can only be spent once, most likely making it pointless.

Anviking replied:
_hoping we can remove the retrying `it` soon_

jonathanknowles replied:
That's interesting! I had no idea that this difference even existed. Thanks for pointing it out.

I'm just thinking how we could make this difference clearer, so that the name of the function makes it obvious whether or not the contents will be retried on failure.

jonathanknowles replied:
Fixed in d4f6fd647456f10c81b993635d71c597a2ea440b.

Anviking reviewed (commented):
Will look closer later, but looks good 👍

jonathanknowles committed:
Make `migrateWallet` endpoint return non-empty tx list.

In the case where there really is nothing to migrate, we already return
a 403 error. So there's no point in using types that make it possible to
return an empty list of transactions as part of a successful response.

jonathanknowles committed:
Resurrect test `MIGRATE_03`.

jonathanknowles committed:
Resurrect test `MIGRATE_04`.

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1427,46 +1458,45 @@ selectAssets ctx (utxo, cp, pending) tx outs transform = do
-                readCheckpoint wid
+            cp <- withExceptT ErrSignPaymentNoSuchWallet
+                $ withNoSuchWallet wid
+                $ readCheckpoint wid
```
Er, what's wrong with the other indentation?

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2310,6 +2413,11 @@ data ErrStartTimeLaterThanEndTime = ErrStartTimeLaterThanEndTime
     } deriving (Show, Eq)
 
+data ErrCreateMigrationPlan
+    = ErrCreateMigrationPlanEmpty
```
Couldn't it just return and empty list of transactions? That seems fairly obvious and easy to handle for users.

jonathanknowles replied:
You're right, it could.

However, the `NothingToMigrate` condition was already a specific error in the old `migrateWallet` API. This PR just reuses the same approach for `createMigrationPlan`.

It covers two cases:
- The wallet is genuinely empty (therefore there really is nothing available to migrate).
- The wallet isn't empty, but the migration algorithm wasn't able to pay for any of the UTxO entries to be migrated, and therefore the plan is empty.

For the second case, I think it's useful to have a specific error here, as we can provide a useful hint to the user:

> I wasn't able to construct a migration plan. This could be because your wallet is empty, or it could be because the amount of ada in your wallet is insufficient to pay for any of the funds to be migrated. Try adding some ada to your wallet before trying again."

For comparison, the **_original_** message was:

>  I can't migrate the wallet with the given ID, because it's either empty or full of small coins which wouldn't be worth migrating."

Anviking replied:
Arguably it might want to be worth including the case-specific `balanceLeftover` in the 403 case. 

Could we have a err403 *but* returning `ApiWalletMigrationPlan` with `selections = []`? It couldn't include an error message though, and perhaps a bit weird, so probably not.

jonathanknowles replied:
> Could we have a err403 but returning ApiWalletMigrationPlan with selections = []? It couldn't include an error message though, and perhaps a bit weird, so probably not.

In that specific case, the `balanceLeftover` would be precisely the same as the wallet's balance. So perhaps it wouldn't be providing any more info than the user already has. (Presumably, they already know their balance, or can easily check it.)

rvl commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1992,26 +2007,141 @@ quitStakePool ctx (ApiT wid) body = do
+    -> MigrationPlan
+    -> Maybe (ApiWalletMigrationPlan n)
+mkApiWalletMigrationPlan s addresses rewardWithdrawal plan
+    | Just selections <- maybeSelections =
```
```case maybeSelections of```

jonathanknowles replied:
Ah yes, the pattern guard is a bit unnecessary here!

(I think it might be a hangover from a previous version of the function.)

I've fixed this in a671d6eb69202f89285f87cad0c034df9a9bf7bf.

rvl reviewed (commented):
This seems pretty good so far :+1:. You could also update the migration endpoint status info in `swagger.yaml`. Unnecessary whitespace changes make it harder to provide a good review.

rvl commented on lib/core-integration/src/Test/Integration/Scenario/API/Byron/Migrations.hs:
```diff
@@ -115,66 +117,259 @@ spec :: forall n.
+            [ expectResponseCode HTTP.status202
+            , expectField
+                (#totalFee . #getQuantity)
+                (`shouldBe` 2_460_400)
```
In integration tests, exact values like this may cause us pain in future.
It might be possible to make the test case more forgiving of changes, without reducing its effectiveness.

jonathanknowles replied:
Hi @rvl. I think I understand your concerns here.

However, I think there are good arguments in favour of using exact values:

- The new migration algorithm is completely deterministic: the plan is fully determined by the current UTxO, with no randomness. So if we start with the same fixture wallet and transaction constraints, we should always end up with precisely the same plan.

- To increase confidence that there are no flaky tests, I've run the entire migration integration test suite 500 times and confirmed that it always passes.

- Stating exact values means we can increase our confidence that the API is producing plans with reasonable fees, since fees are explicitly encoded as literals in the test code. I think it's good to be a bit paranoid here, as the potential cost to a user of miscalculating a fee is that they spend more ada than required (potentially without noticing).

- If someone alters the fixture wallet definitions or the fee calculation algorithm, then it will cause the tests to fail. However, they're likely to fail in an extremely obvious way, and it's likely that we can fix them very quickly just by entering new numbers. It also gives a fresh chance for a human pair of eyes to confirm that fees produced by the API are reasonable.

jonathanknowles committed:
Resurrect test `MIGRATE_05`.

jonathanknowles committed:
Resurrect test `MIGRATE_06`.

jonathanknowles committed:
Resurrect test `MIGRATE_07`.

jonathanknowles committed:
Remove migration test that checks for depletion of source funds.

This check is actually already performed in other integration tests.

In addition, we can easily add this check to the end of tests that don't
have it already.

jonathanknowles committed:
Resurrect test `MIGRATE_08`.

jonathanknowles committed:
Don't retry tests that deplete a fixture wallet's funds.

If such a test fails when run the first time, it's more than likely that
it'll fail when run a second time (as a fixture wallet's funds can only
be spent once).

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2644#discussion_r630565046

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3207,6 +3336,13 @@ instance IsServerError ErrOutputTokenQuantityExceedsLimit where
+    toServerError = \case
+        ErrCreateMigrationPlanEmpty ->
+            -- TODO: Provide a more useful error message:
+            apiError err403 NothingToMigrate "Nothing to migrate"
```
We should provide a more detailed error message here.

At the very least, it needs to explain the types of circumstances that might lead to this error, so the user can take action.

jonathanknowles replied:
Fixed in 713990db02026c286437a7061b275ba41fc72a3e.

jonathanknowles committed:
Use more idiomatic style in `mkApiWalletMigrationPlan`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2644#discussion_r630743466

jonathanknowles committed:
Provide a useful error message when there's nothing to migrate.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2644#discussion_r631650180

jonathanknowles committed:
Remove "disabled" warnings from migration endpoints.

jonathanknowles committed:
Add test `MIGRATE_MULTI_ASSET_01`.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs:
```diff
@@ -114,330 +128,595 @@ spec :: forall n.
+                [ expectResponseCode HTTP.status202
                 , expectField (#totalFee . #getQuantity)
-                    (.> 0)
+                    (`shouldBe` 255_200)
+                , expectField (#selections)
+                    ((`shouldBe` 1) . length)
+                , expectField (#balanceSelected . #ada . #getQuantity)
+                    (`shouldBe` 1_000_000_000_000)
+                , expectField (#balanceLeftover . #ada . #getQuantity)
+                    (`shouldBe` 0)
```
👍

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -17,6 +17,7 @@
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
```
☝️ Guessing your HLS quick-fix added this by mistake

Anviking replied:
Or it was intentional?
<img width="648" alt="Skärmavbild 2021-05-14 kl  18 12 18" src="https://user-images.githubusercontent.com/304423/118298585-ec4b5a00-b4df-11eb-9f95-e7792e2efc72.png">

Could we get by with a `type` alias instead, not to have to disable the warning for the entire module?

jonathanknowles replied:
@Anviking Thanks, this is a good suggestion. 👍🏻 

I also had reservations about disabling this warning. It's a shame that it's not possible to disable these warnings for just a single function. (Or is it?) I notice that we disable this warning in other modules -- perhaps we shouldn't do that.

I agree with you that it's just simpler to use a type synonym here.

I've fixed this in b9136673f1f69610b331a37f9dd50004d85571c7.

Anviking reviewed (approved):
👍

jonathanknowles committed:
Use type synonym for selection results without change.

This allows us to remove the module-wide `fno-warn-redundant-constraints`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2644#discussion_r632640582
